### PR TITLE
sys: make uart_stdio RX optional (attempt #2)

### PR DESCRIPTION
**Pull request by vincent-d**
### Contribution description

This is a rebase of #7974 that fixes conflicts. This allows to not block the low power mode if the shell is not used.

### Issues/PRs references

#7974

### CODE REVIEW

**Comment by jcarrano** [comment]
If I understand this correctly, it is currently impossible to have UART RX compiled in _and_ be able to go to low power mode, right? So if the user wants to use the RX functionality, he has no way to disable it at runtime to be able to go to sleep.

I'd rather not force the user into choosing _either_ RX _or_ low power. It is better to provide a run-time toggle for the RX function, or to bypass the low power block.

From #7947 I get that this is only a problem on chips that don't have a low power capable UART. In those chips, bypassing the mode blockage would mean that there is no input until the MCU wakes up again. In https://github.com/RIOT-OS/RIOT/issues/7947#issuecomment-342533933 you describe a possible solution which I believe is a good one.

Also, I'm trying to find where in the RIOT code is the lpm block, but I cannot. Which CPUs have this limitation?

**Comment by kaspar030** [comment]
> If I understand this correctly, it is currently impossible to have UART RX compiled in _and_ be able to go to low power mode, right?

Nope. On stm32, if an RX callback is configured and uart is initialized (and not powered off through uart_poweroff()), LPM is blocked such that RX still works.
The problem is that stdio_uart always initializes both RX and TX.

**Comment by kaspar030** [comment]
Thanks @vincent-d for taking over!

File: Makefile.dep [hunk 1]
```diff
@@ -396,8 +396,20 @@ ifneq (,$(filter stdio_rtt,$(USEMODULE)))
   USEMODULE += xtimer
 endif
 
-ifneq (,$(filter stdio_uart,$(USEMODULE)))
+ifneq (,$(filter shell,$(USEMODULE)))
+  USEMODULE += stdio_uart_rx
```
Comment (kaspar030) [inline Makefile.dep]: this'll not work with stdio_rtt

File: Makefile.dep [hunk 2]
```diff
@@ -396,8 +396,20 @@ ifneq (,$(filter stdio_rtt,$(USEMODULE)))
...
+endif
+
+ifneq (,$(filter rtt_stdio,$(USEMODULE)))
```
Comment (kaspar030) [inline Makefile.dep]: and this is unrelated?

File: Makefile.dep (hunk 1 above)
Comment (kaspar030) [reply Makefile.dep]: maybe depend on an intermediate module like "stdio_input" and make that depend on stdio_uart_rx if stdio_uart is used?

**Comment by jcarrano** [comment]
> On stm32, [...]

I'm not familiar with STM32, but googling around seems to indicate that it has the ability to be woken up from the UART. If that is the case then it is something that should be addressed by the periph implementation.

> LPM is blocked such that RX still works.

Ok, but what happens if the block is ignored and the device sleeps? Does anything break? Will the RX work after it wakes up again?

If it is OK for a certain application to not have RX, then it does not matter if the answer is yer or no, and ignoring the blockage is enough to fix the issue (or, rather, work around the lack of low power uart in stm32).

If the answer is "yes" (meaning after wake up the RX is still working) then ignoring the blockage, in combination with something like what @vincent-d proposes in https://github.com/RIOT-OS/RIOT/issues/7947#issuecomment-342533933 is good too.

In both cases, _removing_ RX is a bit of an extreme measure.

**Comment by kaspar030** [comment]
> I'm not familiar with STM32, but googling around seems to indicate that it has the ability to be woken up from the UART. If that is the case then it is something that should be addressed by the periph implementation.

This usually loses the first byte, which may or may not be acceptable.

> Ok, but what happens if the block is ignored and the device sleeps? Does anything break? Will the RX work after it wakes up again?

On stm32 (and most platforms I know), yes, unless the MCU sleeps too deep. On stm32, there's e.g., the standby mode, which would lose IO configuration and all RAM contents.

> If it is OK for a certain application to not have RX, then it does not matter if the answer is yer or no, and ignoring the blockage is enough to fix the issue (or, rather, work around the lack of low power uart in stm32).

There's many ways to tackle this issue. As of now, if uart is initialized and powered with an RX callback, RX works, but blocks LPM. As long as we don't have a way to specify that losing bytes is acceptable, it should be that way.

"certain applications" should not initialize uart with RX enabled (e.g., they should *not* pass an rx callback).

This PR tackles the issue by making *stdio* RX optional, by not configuring it trough an RX callback.

More elaborate alternatives would make stdio RX sleep controllable. Maybe make it only active when someone is waiting in a read() call comes to mind.
IMO, until we have that functionality, making this a compile-time option makes sense. This is also nice for output-only applications that want to save the RX buffer.

File: Makefile.dep (hunk 2 above)
Comment (vincent-d) [reply Makefile.dep]: rebase mess I guess ;)

File: Makefile.dep (hunk 1 above)
Comment (vincent-d) [reply Makefile.dep]: just checked for `stdio_uart` being used, is it good enough?

**Comment by aabadie** [comment]
@kaspar030 @jcarrano, I think this is in good shape and already provide an enhancement regarding power management, without the shell. I'm +1 for merging it. What's your opinion (e.g. do you ACK) ?

File: sys/stdio_uart/stdio_uart.c [hunk 3]
```diff
@@ -45,16 +47,28 @@ extern ethos_t ethos;
...
+    cb = NULL;
+    arg = NULL;
+#endif
+
 #ifndef USE_ETHOS_FOR_STDIO
-    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, (uart_rx_cb_t) isrpipe_write_one, &stdio_uart_isrpipe);
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, cb, arg);
 #else
-    uart_init(ETHOS_UART, ETHOS_BAUDRATE, (uart_rx_cb_t) isrpipe_write_one, &stdio_uart_isrpipe);
+    uart_init(ETHOS_UART, ETHOS_BAUDRATE, cb, arg);
```
Comment (kaspar030) [inline sys/stdio_uart/stdio_uart.c]: I think we need to error here. ethos won't work without rx.

**Comment by kaspar030** [comment]
> What's your opinion (e.g. do you ACK) ?

Just gave it a last look. Unfortunately I just realized that as is I think it would break ethos if used with the same uart as stdio (e.g., on the border router or our suit examples).

File: sys/stdio_uart/stdio_uart.c (hunk 3 above)
Comment (vincent-d) [reply sys/stdio_uart/stdio_uart.c]: Is my last commit enough to fix this?

Comment (kaspar030) [reply sys/stdio_uart/stdio_uart.c]: Perfect!

**Comment by kaspar030** [comment]
Please squash!

**Comment by vincent-d** [comment]
Squashed

File: sys/stdio_uart/stdio_uart.c [hunk 4]
```diff
@@ -45,16 +47,31 @@ extern ethos_t ethos;
...
 void stdio_init(void)
 {
+    uart_rx_cb_t cb;
+    void *arg;
+
+#ifdef MODULE_STDIO_UART_RX
+    cb = (uart_rx_cb_t) isrpipe_write_one;
+    arg = &stdio_uart_isrpipe;
+#else
+#ifdef USE_ETHOS_FOR_STDIO
```
Comment (aabadie) [inline sys/stdio_uart/stdio_uart.c]: I'm wondering if we could make stdio_uart_rx a dependency of ethos in the build system instead of preprocessor macros ?

Comment (vincent-d) [reply sys/stdio_uart/stdio_uart.c]: I am not sure about this... @kaspar030 any opinion?

**Comment by aabadie** [comment]
@kaspar030, can you give a final review and eventually ACK this one ?

File: sys/stdio_uart/stdio_uart.c (hunk 4 above)
Comment (kaspar030) [reply sys/stdio_uart/stdio_uart.c]: hm, ethos itself doesn't need *stdio*_uart_rx.
stdio-over-ethos should probably be a pseudomodule...

I guess as is this is fine for now.

**Comment by aabadie** [comment]
Thanks @kaspar030 :)

**Comment by benemorius** [comment]
> > I'm not familiar with STM32, but googling around seems to indicate that it has the ability to be woken up from the UART. If that is the case then it is something that should be addressed by the periph implementation.
> 
> This usually loses the first byte, which may or may not be acceptable.

Is that due to taking too long to wake up and clear the receive buffer or is it actually a documented hardware limitation? I don't think I've ever had to tolerate that problem on any platform, but it's possible I've avoided it by shortening the wakeup sequence.

I recall some time ago moving the main oscillator startup to a later point and making it conditional, but I don't remember if that was for the uart or just to avoid wasting power for xtimer overflow wakeups which come very frequently on EFM32.
